### PR TITLE
ARTEMIS-1856 support delays before deleting addresses & queues

### PR DESCRIPTION
**Pull request by jbertram**

### CODE REVIEW

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java [hunk 1]
```diff
@@ -1540,6 +1561,75 @@ public void run() {
...
+            if (!queue.isInternalQueue() && autoCreated && messageCount == 0 && consumerCount == 0 && consumerRemovedTimestamp != -1) {
+               SimpleString queueName = queue.getName();
+               AddressSettings settings = addressSettingsRepository.getMatch(queue.getAddress().toString());
+               if (settings.isAutoDeleteQueues() && (System.currentTimeMillis() - consumerRemovedTimestamp >= settings.getAutoDeleteQueuesDelay())) {
+                  if (ActiveMQServerLogger.LOGGER.isDebugEnabled()) {
+                     ActiveMQServerLogger.LOGGER.info("deleting auto-created queue \"" + queueName + ".\" consumerCount = " + consumerCount + "; messageCount = " + messageCount + "; isAutoDeleteQueues = " + settings.isAutoDeleteQueues());
+                  }
+
+                  try {
+                     server.destroyQueue(queueName, null, true, false);
```
Comment (sebthom) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java]: What happens if a message was added to the queue in the time between queue.getMessageCount() was invoked and server.destroyQueue() is executed. Maybe ActiveMQServer.destroyQueue() needs to be extended by a "checkMessageCount" parameter?

Comment (michaelandrepearce) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/PostOfficeImpl.java]: The chance of this existed (and still does) in the existing QueueManagerImpl logic where no delay is taken into account.

**Comment by michaelandrepearce** [comment]
LGTM, will merge tomorrow if no one beats me to it.

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd [hunk 2]
```diff
@@ -336,6 +336,14 @@
             </xsd:annotation>
          </xsd:element>
 
+         <xsd:element name="address-queue-scan-period" type="xsd:long" default="30000" maxOccurs="1" minOccurs="0">
```
Comment (michaelandrepearce) [inline artemis-server/src/main/resources/schema/artemis-configuration.xsd]: Can this be added to the artemis-configuration used for testing to keep as much as possible the two in sync

File: artemis-server/src/main/resources/schema/artemis-configuration.xsd [hunk 3]
```diff
@@ -2953,6 +2961,15 @@
                </xsd:annotation>
             </xsd:element>
 
+            <xsd:element name="auto-delete-queues-delay" type="xsd:long" default="0" maxOccurs="1" minOccurs="0">
```
Comment (michaelandrepearce) [inline artemis-server/src/main/resources/schema/artemis-configuration.xsd]: ditto - Can this be added to the artemis-configuration used for testing to keep as much as possible the two in sync

**Comment by michaelandrepearce** [comment]
@jbertram went to merge this morning and during the process and noticed something, have added comment inline

**Comment by michaelandrepearce** [comment]
@jbertram it wasn't a biggie, i guess more important things are a foot, so merging. And can always add the extra config to the other as and when needed in future.

**Pull request closed without merging**